### ISSUE TITLE
no_std support

### ISSUE THREAD

**Issue by MOZGIII** [RustAudio/cpal#347]
I’m investigating the use of cpal in `no_std` environment. I don’t have a particular goal (i.e. embedded device) in mind, but it would be cool to do something with audio in a bare metal environment. Is there a way `cpal` could help me here? Has `no_std` support been looked into before?

**Comment by ishitatsuyuki** [comment]
The point of cpal is to obtain access to the audio API, we barely do any audio processing. Considering embedded platforms should have their own way to do audio I/O, cpal shouldn't be involved and therefore there's no point for no_std.

**Comment by MOZGIII** [comment]
I want to use no_std to make the compiler force me to look into allocations in more detail. I want to build an app that doesn't allocate at all after an initial bootstrap stage.

**Comment by ishitatsuyuki** [comment]
I still don't get your point. What platform are you trying to use cpal on? And what kind of other crates are you going to use?

**Comment by MOZGIII** [comment]
Any of the major platforms. Windows, Linux, macOS, typical.

My point is that I want to cut some capabilities to provide stricter guarantees and make the compiler enforce them.

I'm making a relatively simple app, so I'm planning to implement most of the things manually, or using `core`. I'm planning to make an allocator that has a switch (`AtomicBool`) that will panic when invoked after the initial setup is complete (i.e. buffers are allocated and the streams are configured).

Ideally, I'd like `cpal` to not just be a `no_std`, but also not use `alloc`. This way it'll be guaranteed to do no heap allocations. I assume, `cpal` shouldn't need any allocations after the initial setup, and it'd be in the interest of the `cpal` as well if to have some guarantees of this. By guarantees, I mean that I'd rather have a compile error if I accidentally introduce the code that does allocation, rather than discover it at runtime.

**Comment by ishitatsuyuki** [comment]
If you are reimplementing things anyway, why not also reimplement cpal? The point is, things like no_std really doesn't fit to cpal.

I don't know what exactly cpal would achieve with the allocation-free goal. Sacrificing crate compatibility is a direct consequence of such goals; idiomatic programming is achieved in Rust under the assumption that heap allocation is possible, and again there's no point for cpal to sacrifice that ergonomics to meet a no-allocation requirement.

**Comment by MOZGIII** [comment]
> If you are reimplementing things anyway, why not also reimplement cpal? The point is, things like no_std really doesn't fit to cpal.

I don't want to reimplement cpal, but rather contribute to it, improve it. I disagree that this is not a good fit for `cpal` (as a low-level library) conceptually. It might not be good for `cpal` as it currently is, but it is only because it's not been addressed yet. I understand that it sounds like it's going to be difficult to fit into the design.

I do not see what are you afraid of - sacrificing what ergonomics? If `cpal` does allocations in hot paths (i.e. after the initial setup of the streams) it certainly should not. Could you elaborate on what difficulties you see with the no-allocation requirement?

**Comment by ishitatsuyuki** [comment]
You are right that we should not make allocations in the hot path - but that doesn’t equal that we should have no access to the allocator. The allocator is still useful for things like exception handling or simply reconfiguring streams; you can’t pretend that errors don’t happen.

Also on your idea of making allocations a compiler error: no, Rust is not a good fit for such goals. Rust does not discriminate allocating and non-allocating code except in the very restricted way of no_std-ing the entire crate.

Overall, I don’t see how your proposal can be materialized in cpal. You’re welcomed to file a pull request to satisfy whatever goal you have in mind (eliminating allocation in hot paths?), but only if it does not make the API awkward for the majority of current use cases.

**Comment by MOZGIII** [comment]
I've started playing around with a custom allocator wrapper that can be turned "on" and "off". I made it panic when it's in the "off" state and the `alloc` is called. It's not `no_std`, but it should provide a strong signal if the allocation happens. I think it's going to be useful to write some test code to guarantee the absence of allocations in the hot paths while treating the actual code as a black box.

I get why `no_std` can be unwanted for the whole crate. I've also had some idea in mind around extracting the host drivers into subcrates to better isolate them, and enable `no_std` support conditionally where it makes sense. This work can't be done at once in a single PR with a reasonable level of complexity, nor does it have to really. I'm still figuring out a way to land this though, but either way - what do you think about extracting host-specific subcrates?

**Comment by roderickvd** [comment]
> what do you think about extracting host-specific subcrates?

It's possible. But even then I wonder if you can make those `no_std`. For example, ALSA allocates a `Box<[u8]>` during stream setup once it knows the buffer size. That's outside of the hot path, but would still be in such a sub crate.

I'm closing this for now because I don't see a way forward with this. Feel free to open another issue or PR if you have an idea how to better detect allocations in the audio hot path.

**Comment by MOZGIII** [comment]
`no_std` is actually super easy; doing `no_std` just means you'd explicitly use `alloc` as needed - and then it is clear where/how/if you can expose a no-`alloc` API subset.

The example with the `Box<[u8]>` outside of the hot paths is illustrative. I'd organize my own code into two crates, one for setup that depends on `cpal` with `alloc` feature, and one for all the hot paths that depends on `cpal` with *no* `alloc` feature. Then I can use CI to check that the hot path crate works standalone without depending on any code that needs `alloc`.
